confluent_kafka/schema_registry/schema_registry_client.py

```diff
@@ -18,133 +18,258 @@
+from copy import deepcopy
+from threading import Lock
 
-from requests import Session
+from requests import Session, utils
 
-from .error import SchemaRegistryClientError
-from .schema import SchemaProvider
+from .schema import RegisteredSchema, Schema
+from ..serialization.error import SchemaRegistryError
 
-log = logging.getLogger(__name__)
+# TODO: use six or adopt compat file practicelike requests.
+if sys.version < '3':
+    def _urlencode(value):
+        return urllib.quote(value, safe='')
+else:
+    def _urlencode(value):
+        return urllib.parse.quote(value, safe='')
+
+VALID_AUTH_PROVIDERS = ['URL', 'USER_INFO', 'SASL_INHERIT']
 
-"""
-API Endpoint templates
-"""
-SCHEMAS = 'schemas/ids/{}?format=serialized'
-SUBJECT = 'subjects/{}'
-SUBJECTS = 'subjects'
-VERSION = 'subjects/{}/versions/{}'
-VERSIONS = 'subjects/{}/versions'
-CONFIG = 'config/{}'
-CONFIGS = 'config'
+log = logging.getLogger(__name__)
 
 
-def url_formatter(template, *args):
+class _RestClient(object):
     """
-    Constructs a url string from a template.
+    Manages SchemaRegistryClient HTTP communication
 
     Args:
-        template: url format string
-        args: args to use when resolving url string
-
+        src_conf (dict): Dictionary containing _RestClient configuration
     """
-    return template.format(*args)
+    default_headers = {'Accept': "application/vnd.schemaregistry.v1+json,"
+                                 " application/vnd.schemaregistry+json,"
+                                 " application/json"}
 
+    def __init__(self, src_conf):
+        self.session = Session()
+        conf = self.handle_conf(src_conf)
+        self.base_url = conf['url']
 
-class SchemaRegistryClient(object):
-    """
-    A client that talks to a Schema Registry over HTTP
+    def __del__(self):
+        self.session.close()
 
-    See Also:
-        http://confluent.io/docs/current/schema-registry/docs/intro.html
+    def handle_conf(self, src_conf):
+        """
+        Validates configuration
 
-    Arguments:
-        conf (SchemaRegistryConfig): Schema Registry config object
-    """
+        Args:
+            src_conf (dict): configuration dict
+        """
+        conf = deepcopy(src_conf)
 
-    def __init__(self, conf):
-        self._rest_client = self._RestClient(conf)
-        self.id_to_schema = {}
+        base_url = conf.get('url', '').rstrip('/')
+        if not base_url.startswith('http'):
+            raise ValueError("Invalid URL provided for Schema Registry")
 
-    def __enter__(self):
-        return self
+        ca = conf.pop('ssl.ca.location', None)
+        if ca is not None:
+            self.session.verify = ca
 
-    def __exit__(self, *args):
-        if self._rest_client is not None:
-            self._rest_client.close()
+        key = conf.pop('ssl.key.location', None)
+        cert = conf.pop('ssl.certificate.location', None)
+        if cert is not None and key is not None:
+            self.session = (key, cert)
```

Comment:
Here, you set `self.session` to a tuple, elsewhere you set it to an instance of `Session`. This doesn't seem right.